### PR TITLE
Make COMPASS_ENABLED a reboot-required parameter

### PR DESCRIPTION
**Pull request by peterbarker**
We're crossing too many codepaths without crossing the init() method if this is enabled at runtime as we've been checking to see if the library is *enabled* rather than if it is *initialised*.

Make entry to functions depend on whether the library was initialised rather than whether it was enabled.

Add more places where we check for intiialisation; it is safe to call read if the library isn't initialised - it will simply return false.

As discussed last devcall.

### CODE REVIEW

File: ArduCopter/compassmot.cpp [hunk 1]
```diff
@@ -110,7 +114,7 @@ MAV_RESULT Copter::mavlink_compassmot(const GCS_MAVLINK &gcs_chan)
     }
 
     // get initial compass readings
-    compass.read();
+    bool compass_read_success = compass.read();
```
Comment (WickedShell) [inline ArduCopter/compassmot.cpp:117]: If this fails how can you continue? You won't have a valid `compass_base` presumably?

File: ArduCopter/compassmot.cpp [hunk 2]
```diff
@@ -150,8 +154,9 @@ MAV_RESULT Copter::mavlink_compassmot(const GCS_MAVLINK &gcs_chan)
         motors->set_throttle_passthrough_for_esc_calibration(channel_throttle->get_control_in() / 1000.0f);
         SRV_Channels::push();
 
-        // read some compass values
-        compass.read();
+        // read some compass values; all reads must be successful but
+        // we continue to read the compass even if some have failed.
+        compass_read_success = compass.read() && compass_read_success;
```
Comment (WickedShell) [inline ArduCopter/compassmot.cpp:159]: Again once you fail once aren't you done forever?

File: libraries/AP_Compass/AP_Compass.cpp [hunk 3]
```diff
@@ -1407,13 +1425,16 @@ void Compass::try_set_initial_location()
 bool
 Compass::use_for_yaw(void) const
 {
-    return healthy(0) && use_for_yaw(0);
+    return _initialised && healthy(0) && use_for_yaw(0);
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass.cpp:1428]: check enabled param here

**Comment by tridge** [comment]
need a way to test compass failover, perhaps with RCn_OPTION like INS and GPS

**Pull request closed without merging**